### DISCUSSION TITLE
parallel/unroll/vectorize的区别？

### DISCUSSION THREAD

**Discussion by hbsun2113** [mlc-ai/mlc-zh#82] [Q&A]
我不确定是否是自己遗漏了，虽然在课程或者作业中频繁提到这3个操作parallel/unroll/vectorize，但我没有找到对其的解释和区分(e.g. 使用场景)。在TVM文档中也没有找到解释。请问可以帮忙分辨这三者之间的不同吗？
![image](https://user-images.githubusercontent.com/27994442/181756529-9ddc0100-35a8-48c7-a5f9-53ebbbbb3f08.png)

我自己粗浅的理解:
1. parallel是同时执行多条语句。
1. vectorize是SIMD，即使用一条指令操纵多条数据。
2. unroll是循环展开，减少了循环中的分支判断开销。同时也属于增加并行度的一种方法？

**Comment by Hzfengsy** [comment]
1. Parallel相当于把循环底下的执行部分做成了一个线程，提供线程级并行
2. vectorize是SIMD没错
3. unroll是循环展开（传统编译器常见优化），但不会增加并行度

**Comment by yzh119** [comment]
@hbsun2113 关于问题三，“并行度”这个词并没有一个准确的定义，如果你指的是IPC(Instruction Per Cycle)，那么你是对的，loop unroll是经典的提高IPC的方法之一。

**Reply by hbsun2113** [reply]
请问unroll是通过什么方式增加IPC的？是因为减少了分支判断的开销还是因为消除了语句之间的顺序关系？

**Reply by hbsun2113** [reply]
unroll 用于将循环展开，即如果循环次数是可以在编译时确定的话，编译器将会把带有判断和跳转的循环代码展开成串行代码。这样做的好处主要是减少了判断语句，此外还有利于编译器发现数据依赖从而更好地分配寄存器。缺点是可能会增加寄存器的使用，有潜在的降低GPU占用率的风险。
[link](https://openmlsys.github.io/chapter_accelerator/accelerator_practise.html#:~:text=unroll%20%E7%94%A8%E4%BA%8E%E5%B0%86%E5%BE%AA%E7%8E%AF%E5%B1%95%E5%BC%80%EF%BC%8C%E5%8D%B3%E5%A6%82%E6%9E%9C%E5%BE%AA%E7%8E%AF%E6%AC%A1%E6%95%B0%E6%98%AF%E5%8F%AF%E4%BB%A5%E5%9C%A8%E7%BC%96%E8%AF%91%E6%97%B6%E7%A1%AE%E5%AE%9A%E7%9A%84%E8%AF%9D%EF%BC%8C%E7%BC%96%E8%AF%91%E5%99%A8%E5%B0%86%E4%BC%9A%E6%8A%8A%E5%B8%A6%E6%9C%89%E5%88%A4%E6%96%AD%E5%92%8C%E8%B7%B3%E8%BD%AC%E7%9A%84%E5%BE%AA%E7%8E%AF%E4%BB%A3%E7%A0%81%E5%B1%95%E5%BC%80%E6%88%90%E4%B8%B2%E8%A1%8C%E4%BB%A3%E7%A0%81%E3%80%82%E8%BF%99%E6%A0%B7%E5%81%9A%E7%9A%84%E5%A5%BD%E5%A4%84%E4%B8%BB%E8%A6%81%E6%98%AF%E5%87%8F%E5%B0%91%E4%BA%86%E5%88%A4%E6%96%AD%E8%AF%AD%E5%8F%A5%EF%BC%8C%E6%AD%A4%E5%A4%96%E8%BF%98%E6%9C%89%E5%88%A9%E4%BA%8E%E7%BC%96%E8%AF%91%E5%99%A8%E5%8F%91%E7%8E%B0%E6%95%B0%E6%8D%AE%E4%BE%9D%E8%B5%96%E4%BB%8E%E8%80%8C%E6%9B%B4%E5%A5%BD%E5%9C%B0%E5%88%86%E9%85%8D%E5%AF%84%E5%AD%98%E5%99%A8%E3%80%82%E7%BC%BA%E7%82%B9%E6%98%AF%E5%8F%AF%E8%83%BD%E4%BC%9A%E5%A2%9E%E5%8A%A0%E5%AF%84%E5%AD%98%E5%99%A8%E7%9A%84%E4%BD%BF%E7%94%A8%EF%BC%8C%E6%9C%89%E6%BD%9C%E5%9C%A8%E7%9A%84%E9%99%8D%E4%BD%8EGPU%E5%8D%A0%E7%94%A8%E7%8E%87%E7%9A%84%E9%A3%8E%E9%99%A9%E3%80%82)

**Reply by yzh119** [reply]
sorry 之前没有看到你的问题，你贴的这段话基本上是正确的。
在你看的这个例子中，#pragma unroll提示编译器去展开一个循环，其中unroll factor是由nvcc决定的，在tvm中执行unroll循环会显式完全展开。

**Comment by hbsun2113** [comment]
补充一点，我发现parallel、unroll、vectorize的顺序一般是固定的（逐级包含：parallel最外层，unroll在中间、vectorize最内层），我猜测是因为它们粒度逐渐变小：parallel(线程级)、unroll(语句级)、vectorize(指令级)。
![image](https://user-images.githubusercontent.com/27994442/181915114-d714cd72-3403-42d7-a8cf-c6c63571638b.png)

**Reply by hbsun2113** [reply]
@Hzfengsy @yzh119 我对这类优化知识缺乏，请问有推荐的学习资料吗？

**Reply by yzh119** [reply]
CAAQA(Computer Architecture: A Quantative Approach)的第3.2节
